x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/enrollment/CreateEnrollmentToken.java

```diff
@@ -0,0 +1,215 @@
+import org.elasticsearch.xpack.security.tool.CommandLineHttpClient;
+import org.elasticsearch.xpack.security.tool.HttpResponse;
+
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.nio.charset.StandardCharsets;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class CreateEnrollmentToken {
+    protected static final String ENROLL_API_KEY_EXPIRATION = "30m";
+
+    private static final Logger logger = LogManager.getLogger(CreateEnrollmentToken.class);
+    private final Environment environment;
+    private final SSLService sslService;
+    private final CommandLineHttpClient client;
+
+    public CreateEnrollmentToken(Environment environment) {
+        this(environment, new CommandLineHttpClient(environment));
+    }
+
+    // protected for testing
+    protected CreateEnrollmentToken(Environment environment, CommandLineHttpClient client) {
+
+        this.environment = environment;
+        this.sslService = new SSLService(environment);
+        this.client = client;
+    }
+
+    public String create(Terminal terminal, String user, SecureString password) throws Exception {
+        if (XPackSettings.ENROLLMENT_ENABLED.get(environment.settings()) != true) {
+            throw new UserException(ExitCodes.CONFIG,
+                "'xpack.security.enrollment' must be enabled to create an enrollment token");
+        }
+
+        final KeyConfig keyConfig = sslService.getHttpTransportSSLConfiguration().keyConfig();
+        if (keyConfig instanceof StoreKeyConfig == false) {
+            throw new UserException(ExitCodes.CONFIG,
+                "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration is not configured with a keystore");
+        }
+        final List<Tuple<PrivateKey, X509Certificate>> httpCaKeysAndCertificates =
+            ((StoreKeyConfig) keyConfig).getPrivateKeyEntries(environment).stream()
+                .filter(t -> t.v2().getBasicConstraints() != -1).collect(Collectors.toList());
+        if (httpCaKeysAndCertificates.isEmpty()) {
+            throw new UserException(ExitCodes.CONFIG,
+                "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore doesn't contain any " +
+                    "PrivateKey entries where the associated certificate is a CA certificate");
+        } else if (httpCaKeysAndCertificates.size() > 1) {
+            throw new UserException(ExitCodes.CONFIG,
+                "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore contain multiple " +
```

Comment:
```suggestion
                "Unable to create an enrollment token. Elasticsearch node HTTP layer SSL configuration Keystore contains multiple " +
```

Reply:
Should we also point to the setting key in the error message ? As in "which is the HTTP layer SSL configuration Keystore" ?